File: my-folder/0801-is-graph-bipartite/solution.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> color(n,-1);
        queue<int> q;
        
        for (int node = 0; node<n; node++)
        {
            if ( color[node] == -1) 
            {
                color[node] = 0;
                q.push(node);
                while (!q.empty())
                {
                    int curr = q.front();
                    q.pop();
                    for (auto nei:graph[curr] )
                    {
                        if ( color[nei]==-1 )
                        {
                            color[nei] = 1 - color[curr];
                            q.push(nei);
                        }
                        else if ( color[nei] == color[curr] )
                        {
                            return false;
                        }

                    }
                }

            }
        }
        return true;
    }
};

File: my-folder/0801-is-graph-bipartite/solution.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        
        n = len(graph)
        color = [-1]*n
        q = []
        for node in range(n):
            if color[node] == -1: 
                color[node] = 0
                q.append(node)
                while q:
                    curr = q.pop(0)
                    for nei in graph[curr]:
                        if color[nei] == -1:
                            color[nei] = 1 - color[curr]
                            q.append(nei)
                        elif color[nei] == color[curr]:
                            return False
        return True

